### ISSUE TITLE
Proposal: `CACHE` (extended Mount-based caching)

### ISSUE THREAD

**Issue by adamgordonbell** [earthly/earthly#1399]
## Problem Statement:

Some build tools are computationally expensive when run initially but quick on subsequent runs because they work incrementally. Earthly's layer-based caching ensures these tools are called no more than necessary, but, at present, when they are called the incremental features of these tools are lost. 

**Related Issues:**
* Earthly: #1139
* Earthly: #1129
* Earthly: #788
* Buildkit: https://github.com/moby/buildkit/issues/1512
* Buildkit: https://github.com/moby/buildkit/issues/1723
* Buildkit: https://github.com/moby/buildkit/issues/1545

**Maven Example:**
Maven’s `mvn install` is slow to run the first time because all packages are downloaded to a cache. If you add a new package, it will only need to download that one new package. However, in earthly, whenever a new package is added, `mvn install` will need to redownload every package.

**Dialyzer Example:**
Elixir’s Dialyzer PLT files are expensive to generate the first time, often taking 20 minutes to generate from scratch. But, subsequent runs take seconds because they use the output of the previous runs to prevent rework.

**Incremental Compilers:**
Some compilers compile slowly initially, but subsequent compiles use the intermediate files from the previous run to gain speed. With Earthly, every compile is the first run. 

An ideal solution would have the following attributes:

- Not harm repeatability. 
- Be easily understandable
- Work with existing cache options (layer cache, remote inline cache, remote explicit cache)

## Proposed Solution:

Introduce a new keyword `CACHE` which can be used inside a target like so:

```
target:
      FROM alpine
      CACHE /path/to/cache
      RUN ...
      RUN ...
```

`CACHE` mounts a shared volume to that path that will be available in all subsequent RUNs within that target. This volume will be reused across runs. With no additional flags provided this works identically adding a [mount-based caching](https://docs.earthly.dev/docs/guides/advanced-local-caching#option-2-mount-based-caching-advanced) statement to every run in a target.

Flags will extend this behavior further

## Cache Persistence

By default, volumes used for caching are not part of the resulting image. This can be a problem if you need the cached file in subsequent build targets. Cache persistence overcomes this by also copying the contents of the cache in the image itself at the end of the target as an additional layer:

```
one:
      FROM alpine
      CACHE /path/to/cache
      RUN touch /path/to/cache/one
      # at this point earthly copies the contents of the cache into /path/to/cache in the image,
      # which makes it available to anyone inheriting this image (or if the image is pushed to a registry)
two:
      from +one
      # cached file is available here, thanks to --persist
      RUN ls /path/to/cache/one
```

The behavior is similar to this Buildkit proposal: https://github.com/moby/buildkit/issues/1545

Open Questions:
* Given that this type of cache could potentially accumulate data forever, what's a good way for it to be GC'd either manually or automatically?
  * Possible solution: `CACHE --gc="30 days" --persist /path/to/cache`. This would tell Earthly to wipe this cache clean automatically every 30 days.

These questions can be answered at a later time, for now we will leave it to the user to reset their cache using `--no-cache` or `earthly prune --reset` if they need to.

## Share Remote Cache (Phase II)

A Phase II of this feature may be implemented in the future, as described in https://github.com/earthly/earthly/issues/1540.

**Comment by andrewsykim** [comment]
> Should --persist be on by default?
> Is there a more evocative name than persist?

Regarding the `--persist` flag, I wonder if it's worth making a distinction between these two features:
1. Multi-target shared cache mounts (for compile optimizations like Maven)
2. Cache mount becoming part of the image fileystem (for runtime optimizations like node_modules)

I would think maybe 1) should be the default behavior of `CACHE` since you can already achieve single target cache mounts with `RUN --mount=type=cache`. 2) might warrant a new flag like `--include-image` or `--runtime` to indicate the cache gets embedded into the layer's filesystem.

**Comment by brandonSc** [comment]
Updated to describe persistent cache without `--persist` flag (i.e. cache always persists, there's no need to use a flag).

Also moved Phase II implementation into a separate issue: https://github.com/earthly/earthly/issues/1540 since Phase I described here will be released first.

**Comment by vladaionescu** [comment]
This has been released as experimental in Earthly v0.6.3. 🎉 

To enable this feature you need to use `VERSION --use-cache-command 0.6`. Some usage examples are available in [examples/cache-command](https://github.com/earthly/earthly/tree/main/examples/cache-command) of this repo.

**Comment by mwopfner** [comment]
I would like to have the possibility to disable the persistence of the cache into the container image. Maybe with `--no-persist`.
My Use Case is as follow:
We do some expensive work (e.g. compiling large C++ application) so we want to cache the _build_ folder.
Later on we want to get some files out of the _build_ folder in a custom UDC. The _build_ folder has multiple GB which will now be persisted into a layer, but we only need the _build_ folder in the _+build_ target.
The option with `RUN --mount=type=cache` does now work, because the UDC will not have access to it.
~~~
build:
    CACHE ./build
    RUN ...  # Do the build
    DO +MY_UDC    #The UDC operates on the cached build folder

foo:
   FROM +build
   # We do not need ./build here
~~~
So in short: the _build_ folder should be cached for all commands in _+build_ but not for any other targets.

**Comment by brad-jones** [comment]
Can not copy files into the cache volume. Hopefully the following demonstrates my issue.

```
VERSION --use-cache-command 0.6
FROM busybox:latest

scripts-not-copied-to-cache-volume:
  CACHE /scripts
  COPY ./scripts/. /scripts/.
  RUN ls -hal /scripts # empty dir

scripts-touched-inside-cache-volume:
  CACHE /scripts
  RUN touch /scripts/foo.sh
  RUN ls -hal /scripts # dir contains "foo.sh"
```

FWIW I want to do this because I want to run a script as part of the build but not commit that script into the final image.
Sure I could inline the script into the Earthfile but that gets messy with larger scripts. Or I could use the _"builder"_ pattern but sometimes it's difficult to identify all the changes that some operation makes, in order to save them into the ARTIFACT stage.

What I really want is the ability to `RUN --mount,type=bind ...` as per https://github.com/earthly/earthly/issues/1139

**Comment by jimmykarily** [comment]
We make heavy use of [`qemu` in the Kairos project](https://github.com/kairos-io/kairos/blob/master/Earthfile#L477) for end to end testing. Those tests could really benefit from having `/dev/kvm` available from the host. We have tests that jump from a couple of minutes to 25 minutes just because of this. I guess other special devices would also be needed in other cases.

**Comment by mwopfner** [comment]
@jimmykarily I think #1139 is more like the feature you need to mount `/dev/*` devices.

**Comment by vladaionescu** [comment]
Closing this as complete. Some details might have not been completed, but it's better to open new issues instead.
